File: Search/BinarySearch.exs
defmodule BinarySearch do
  def search(list, key) do
    if list != Enum.sort(list) do
      raise ArgumentError, message: "list:"
    end

    do_search(Enum.with_index(list), key)
  end

  defp do_search([], _key), do: :not_found

  defp do_search(list, key) do
    middle = length(list) / 2 |> round

    {left, right} = Enum.split(list, middle)
    {item, index} = List.last(left)
    left = Enum.drop(left, -1)

    cond do
      key == item -> {:ok, index}
      key < item  -> do_search(left, key)
      key > item  -> do_search(right, key)
    end
  end
end

File: Search/Haskell/binarySearch.hs
binarySearch :: [Int] -> Int -> Int
binarySearch list element = 
    binarySearch' list 0 (size - 1) element
    where size = length list

binarySearch' :: [Int] -> Int -> Int -> Int -> Int
binarySearch' list left right element
    | right < left = -1
    | (list !! middle) == element = middle
    | (list !! middle) < element = binarySearch' list (middle + 1) right element
    | otherwise = binarySearch' list left (middle - 1) element
    where middle = (left + right) `div` 2